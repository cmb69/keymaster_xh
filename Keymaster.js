/**
 * The Keymaster class.
 *
 * @file
 *
 * @version   SVN: $Id$
 * @author    Christoph M. Becker <cmbecker69@gmx.de>
 * @copyright 2013 Christoph M. Becker <http://3-magi.net/>
 * @license   http://www.gnu.org/licenses/gpl.html GNU GPLv3
 */

/**
 * The Keymaster class.
 *
 * @constructor
 *
 */
Keymaster = function () {
    /**
     * The warning dialog.
     *
     * @type {HTMLDivElement}
     *
     * @private
     */
    this._warningDialog = null;

    /**
     * The warning text.
     *
     * @type {Text}
     *
     * @private
     */
    this._warningText = '';

    /**
     * The timer handle.
     *
     * @type {Number}
     *
     * @private
     */
    this._timer = null;

    this.createWarningDialog();
    this._timer = setInterval(this.bind(this.requestRemainingTime),
            Keymaster.config.pollInterval);
};

/**
 * The configuration.
 *
 * @type {Object}
 *
 * @static
 */
Keymaster.config = null;

/**
 * The localization
 *
 * @type {Object}
 *
 * @static
 */
Keymaster.l10n = null;

/**
 * Creates the warning dialog.
 *
 * @method
 *
 * @protected
 */
Keymaster.prototype.createWarningDialog = function () {
    var dialog, message, para, text, div, button;

    dialog = document.createElement("div");
    dialog.id = "keymaster";
    dialog.style.display = "none";
    message = document.createElement("div");
    message.className = "keymaster_message";
    para = document.createElement("p");
    text = document.createTextNode("");
    this._warningText = text;
    para.appendChild(text);
    message.appendChild(para);
    div = document.createElement("div");
    div.className = "keymaster_buttons";
    button = document.createElement("button");
    button.onclick = this.bind(this.resetSession);
    text = document.createTextNode(Keymaster.l10n.button);
    button.appendChild(text);
    div.appendChild(button);
    message.appendChild(div);
    dialog.appendChild(message);
    document.body.appendChild(dialog);
    this._warningDialog = dialog;
};

/**
 * Shows the warning dialog.
 *
 * @method
 *
 * @protected
 */
Keymaster.prototype.showWarningDialog = function () {
    this._warningDialog.style.display = "block";
};

/**
 * Hides the warning dialog.
 *
 * @method
 *
 * @protected
 */
Keymaster.prototype.hideWarningDialog = function () {
    this._warningDialog.style.display = "none";
};

/**
 * Requests the remaining session time from the server.
 *
 * @method
 *
 * @protected
 */
Keymaster.prototype.requestRemainingTime = function () {
    var request = new XMLHttpRequest();
    var url = "./?&keymaster_time";

    request.open("GET", url);
    // The following should simply use Function.prototype.bind() or a fallback,
    // but IE 8 doesn't set window.event, so we'd loose the reference to
    // the request object.
    request.onreadystatechange = (function (that) {
        return function () {
            that.receiveRemainingTime(request);
        };
    }(this));
    request.send(null);
};

/**
 * Handles responses for the requested session time.
 *
 * @method
 *
 * @param {XMLHttpRequest} request A request object.
 *
 * @protected
 */
Keymaster.prototype.receiveRemainingTime = function (request) {
    if (request.readyState == 4 && request.status == 200) {
        var remaining = +request.responseText;
        if (remaining != remaining) {
            clearInterval(this._timer);
            if (confirm(Keymaster.l10n.error)) {
                open(url);
            }
        } else if (remaining < 0) {
            location.href = location.href;
        } else if (remaining == 0) {
            location.href = "./?&logout";
        } else if (remaining < Keymaster.config.warn) {
            this.updateWarning(remaining);
            this.showWarningDialog();
        } else {
            this.hideWarningDialog();
        }
    }
};

/**
 * Resets the session timeout.
 *
 * @method
 *
 * @protected
 */
Keymaster.prototype.resetSession = function () {
    var request = new XMLHttpRequest();

    request.open("POST", "./");
    request.setRequestHeader("Content-Type",
            "application/x-www-form-urlencoded");
    request.send("keymaster_reset");
    this.hideWarningDialog();
};

/**
 * Updates the warning message.
 *
 * @method
 *
 * @param {Number} seconds The remaining time in seconds.
 *
 * @protected
 */
Keymaster.prototype.updateWarning = function (seconds) {
    var min = Math.ceil(seconds / 60);

    if (min == 1) {
        text = Keymaster.l10n.warning_singular;
    } else if (min >= 2 && min <= 4) {
        text = Keymaster.l10n.warning_paucal;
    } else {
        text = Keymaster.l10n.warning_plural;
    }
    this._warningText.nodeValue = text.replace(/{{{TIME}}}/, min);
};

/**
 * Creates a new function bound to this. Provides a workaround for possibly
 * missing Function.prototype.bind().
 *
 * @method
 *
 * @param {Function} func A function to bind.
 *
 * @returns {Function}
 *
 * @protected
 */
Keymaster.prototype.bind = function (func) {
    var that = this;

    return function () {
        func.apply(that, arguments);
    };
};

/*
 * Register Keymaster instantiation on load.
 */
if (typeof addEventListener != "undefined") {
    addEventListener("load", function () {
        new Keymaster();
    });
} else if (typeof attachEvent != "undefined") {
    attachEvent("onload", function () {
        new Keymaster();
    });
}
