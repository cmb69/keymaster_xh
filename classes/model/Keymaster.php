<?php

/**
 * Copyright (C) 2013-2023 Christoph M. Becker
 *
 * This file is part of Keymaster_XH.
 *
 * Keymaster_XH is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * Keymaster_XH is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with Keymaster_XH.  If not, see <http://www.gnu.org/licenses/>.
 */

namespace Keymaster\Model;

use Plib\Document;
use Plib\DocumentStore;

final class Keymaster implements Document
{
    /** @var string */
    private $key;

    /** @var int */
    private $timestamp;

    public static function updateIn(DocumentStore $store): self
    {
        $that = $store->update("key", self::class);
        assert($that instanceof self);
        return $that;
    }

    public static function fromString(string $contents, string $key): self
    {
        $record = explode(",", rtrim($contents), 2);
        return new self($record[0], count($record) === 2 ? (int) $record[1] : 0);
    }

    public function __construct(string $key, int $timestamp)
    {
        $this->key = $key;
        $this->timestamp = $timestamp;
    }

    public function toString(): string
    {
        return $this->key . "," . $this->timestamp;
    }

    private function hasKey(): bool
    {
        return $this->key !== "";
    }

    public function acceptKey(string $key, int $now, int $duration): bool
    {
        if ($key === $this->key) {
            $this->timestamp = $now;
            return true;
        }
        if ($this->isFree($now, $duration)) {
            $this->key = $key;
            $this->timestamp = $now;
            return true;
        }
        return false;
    }

    public function grantKey(callable $genKey, int $now, int $duration): ?string
    {
        if (!$this->isFree($now, $duration)) {
            return null;
        }
        $this->key = $genKey();
        return $this->key;
    }

    public function revokeKey(string $key): void
    {
        if ($this->key === $key) {
            $this->key = "";
            $this->timestamp = 0;
        }
    }

    public function isFree(int $now, int $duration): bool
    {
        return !$this->hasKey() || $now - $this->timestamp > $duration;
    }

    public function set(string $key, int $timestamp): void
    {
        $this->key = $key;
        $this->timestamp = $timestamp;
    }
}
