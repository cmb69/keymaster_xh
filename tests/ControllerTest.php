<?php

/**
 * Copyright (C) 2023 Christoph M. Becker
 *
 * This file is part of Keymaster_XH.
 *
 * Keymaster_XH is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * Keymaster_XH is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with Keymaster_XH.  If not, see <http://www.gnu.org/licenses/>.
 */

namespace Keymaster;

use ApprovalTests\Approvals;
use Keymaster\Infra\Model;
use Keymaster\Infra\Request;
use Keymaster\Infra\View;
use PHPUnit\Framework\TestCase;

class ControllerTest extends TestCase
{
    public function testDeniesLoginIfNotFree(): void
    {
        $model = $this->createMock(Model::class);
        $model->method("isFree")->willReturn(false);
        $sut = new Controller($model, $this->view());

        $request = $this->createStub(Request::class);
        $request->method("wantsLogin")->willReturn(true);
        $response = $sut($request);

        $this->assertTrue($response->clearedF());
        Approvals::verifyHtml($response->output());
    }

    public function testLoginSucceedsIfFreeAndKeyIsGiven(): void
    {
        $model = $this->createMock(Model::class);
        $model->method("isFree")->willReturn(true);
        $model->expects($this->once())->method("give")->willReturn(true);
        $sut = new Controller($model, $this->view());

        $request = $this->createStub(Request::class);
        $request->method("isLogin")->willReturn(true);
        $response = $sut($request);

        $this->assertEquals("", $response->output());
    }

    public function testLoginWarnsIfKeyIsNotGivenEvenIfFree(): void
    {
        $model = $this->createMock(Model::class);
        $model->method("isFree")->willReturn(true);
        $model->expects($this->once())->method("give")->willReturn(false);
        $model->method("filename")->willReturn("./plugins/keymaster/key");
        $sut = new Controller($model, $this->view());

        $request = $this->createStub(Request::class);
        $request->method("isLogin")->willReturn(true);
        $response = $sut($request);

        Approvals::verifyHtml($response->output());
    }

    public function testLoginFailsIfNotFree(): void
    {
        $model = $this->createMock(Model::class);
        $model->method("isFree")->willReturn(false);
        $sut = new Controller($model, $this->view());

        $request = $this->createStub(Request::class);
        $request->method("isLogin")->willReturn(true);
        $request->method("su")->willReturn("Current_Page");
        $response = $sut($request);

        $this->assertTrue($response->logout());
        $this->assertEquals("http://example.com/?Current_Page", $response->location());
    }

    public function testLogoutSucceedsIfKeyCanBeTaken(): void
    {
        $model = $this->createMock(Model::class);
        $model->expects($this->once())->method("take")->willReturn(true);
        $sut = new Controller($model, $this->view());

        $request = $this->createStub(Request::class);
        $request->method("isLogout")->willReturn(true);
        $response = $sut($request);

        $this->assertEquals("", $response->output());
    }

    public function testLogoutWarnsIfKeyCannotBeTaken(): void
    {
        $model = $this->createMock(Model::class);
        $model->expects($this->once())->method("take")->willReturn(false);
        $model->method("filename")->willReturn("./plugins/keymaster/key");
        $sut = new Controller($model, $this->view());

        $request = $this->createStub(Request::class);
        $request->method("isLogout")->willReturn(true);
        $response = $sut($request);

        Approvals::verifyHtml($response->output());
    }

    public function testAnswersRemainingTimeIfRequestedByAdmin(): void
    {
        $model = $this->createMock(Model::class);
        $model->method("remainingTime")->willReturn(600);
        $sut = new Controller($model, $this->view());

        $request = $this->createStub(Request::class);
        $request->method("isTimeRequested")->willReturn(true);
        $request->method("isAdmin")->willReturn(true);
        $response = $sut($request);

        $this->assertEquals("text/plain", $response->contentType());
        $this->assertEquals("600", $response->output());
    }

    public function testAnswersThatNoTimeIsRemainingIfRequestedByVisitor(): void
    {
        $model = $this->createMock(Model::class);
        $sut = new Controller($model, $this->view());

        $request = $this->createStub(Request::class);
        $request->method("isTimeRequested")->willReturn(true);
        $request->method("isAdmin")->willReturn(false);
        $response = $sut($request);

        $this->assertEquals("text/plain", $response->contentType());
        $this->assertEquals("-1", $response->output());
    }


    public function testResetsTheTimeIfRequestedByAdmin(): void
    {
        $model = $this->createMock(Model::class);
        $model->expects($this->once())->method("reset")->willReturn(true);
        $sut = new Controller($model, $this->view());

        $request = $this->createStub(Request::class);
        $request->method("isAdmin")->willReturn(true);
        $request->method("isResetRequested")->willReturn(true);
        $response = $sut($request);

        $this->assertEquals("text/plain", $response->contentType());
        $this->assertEquals("1", $response->output());
    }

    public function testResetsTheTimeOnAdminRequestIfSessionIsNotExpired(): void
    {
        $model = $this->createMock(Model::class);
        $model->method("sessionHasExpired")->willReturn(false);
        $model->expects($this->once())->method("reset")->willReturn(true);
        $sut = new Controller($model, $this->view());

        $request = $this->createStub(Request::class);
        $request->method("isAdmin")->willReturn(true);
        $response = $sut($request);

        $this->assertEquals("", $response->output());
    }

    public function testWarnsIfResettingTimeFailsEvenIfSessionIsNotExpired(): void
    {
        $model = $this->createMock(Model::class);
        $model->method("sessionHasExpired")->willReturn(false);
        $model->expects($this->once())->method("reset")->willReturn(false);
        $model->method("filename")->willReturn("./plugins/keymaster/key");
        $sut = new Controller($model, $this->view());

        $request = $this->createStub(Request::class);
        $request->method("isAdmin")->willReturn(true);
        $response = $sut($request);

        Approvals::verifyHtml($response->output());
    }

    public function testLogsOutOnAdminRequestIfSessionIsExpired(): void
    {
        $model = $this->createMock(Model::class);
        $model->method("sessionHasExpired")->willReturn(true);
        $sut = new Controller($model, $this->view());

        $request = $this->createStub(Request::class);
        $request->method("isAdmin")->willReturn(true);
        $request->method("su")->willReturn("Current_Page");
        $response = $sut($request);

        $this->assertTrue($response->logout());
        $this->assertEquals("http://example.com/?Current_Page", $response->location());
    }

    public function testReturnsEmptyResponseOtherwise(): void
    {
        $model = $this->createMock(Model::class);
        $sut = new Controller($model, $this->view());

        $request = $this->createStub(Request::class);
        $response = $sut($request);

        $this->assertEquals("", $response->output());
    }

    private function view(): View
    {
        return new View("./templates/", XH_includeVar("./languages/en.php", "plugin_tx")["keymaster"]);
    }
}
