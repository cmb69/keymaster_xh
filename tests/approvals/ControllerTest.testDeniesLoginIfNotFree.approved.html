<p class="xh_fail">Somebody else is editing the site now. Try again later.</p>
